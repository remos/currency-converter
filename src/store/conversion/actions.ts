import { createAction } from '@reduxjs/toolkit';
import { ConversionsMap, CurrencyCode, Holding } from 'types';

export const recalculateConversion = createAction<{
  base: Holding;
  term: CurrencyCode;
  conversions: ConversionsMap;
}>('recalculateConversion');

export const clearConversion = createAction('clearConversion');
