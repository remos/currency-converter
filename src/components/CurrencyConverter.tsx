import React, { useState } from 'react';
import { ConversionsMap, CurrenciesMap, CurrencyCode, Holding, RootState } from 'types';
import { getDecimals, currencies as defaultCurrencies, conversionMap } from '../data';
import CurrencySelector from './CurrencySelector';
import ConversionTrail from './ConversionTrail';
import { useSelector } from 'react-redux';
import { dispatch } from '../store';
import { clearConversion, recalculateConversion } from '../store/conversion/actions';
import styled from 'styled-components';

const triggerRecalculateIfNeeded = (
  baseCurrency: CurrencyCode,
  baseAmount: number,
  term: CurrencyCode,
  conversions: ConversionsMap
) => {
  if (!baseCurrency || isNaN(baseAmount) || !term) {
    dispatch(clearConversion());
  } else {
    dispatch(
      recalculateConversion({
        base: {
          currency: baseCurrency,
          amount: baseAmount,
        },
        term,
        conversions,
      })
    );
  }
};

const Wrapper = styled.div`
  background-color: white;
  padding: 20px;
  border-radius: 4px;
  border: 1px solid #aaa;
`;

const InputWrapper = styled.div`
  margin: 1rem;
`;

const ConversionTrailWrapper = styled.div`
  margin-top: 1rem;
`;

const Error = styled.span`
  color: red;
`;

const CurrencyConverter: React.FC<{
  conversions: ConversionsMap;
  currencies: CurrenciesMap;
}> = ({ conversions = conversionMap, currencies = defaultCurrencies }) => {
  const error: string = useSelector<RootState, string>((state) => state.conversion.error);
  const holdings: Holding[] = useSelector<RootState, Holding[]>(
    (state) => state.conversion.holdings
  );

  const [baseValue, setBaseValue] = useState(0);
  const [baseCurrency, setBaseCurrency] = useState<CurrencyCode>();
  const [termCurrency, setTermCurrency] = useState<CurrencyCode>();

  const decimals = getDecimals(baseCurrency, currencies);
  const step = isNaN(decimals) ? 1 : 10 ** -decimals;

  return (
    <Wrapper>
      <InputWrapper>
        <input
          type="number"
          value={baseValue}
          step={step}
          min={0}
          onChange={(e) => {
            const amount = parseFloat(e.target.value);
            setBaseValue(amount);
            triggerRecalculateIfNeeded(baseCurrency, amount, termCurrency, conversions);
          }}
        />
        <CurrencySelector
          currencies={currencies}
          value={baseCurrency}
          onChange={(e) => {
            setBaseCurrency(e.target.value);
            triggerRecalculateIfNeeded(
              e.target.value,
              baseValue,
              termCurrency,
              conversions
            );
          }}
        />
        <> to </>
        <CurrencySelector
          currencies={currencies}
          value={termCurrency}
          onChange={(e) => {
            setTermCurrency(e.target.value);
            triggerRecalculateIfNeeded(
              baseCurrency,
              baseValue,
              e.target.value,
              conversions
            );
          }}
        />
      </InputWrapper>
      {error ? <Error>{error}</Error> : null}
      {holdings && holdings.length ? (
        <ConversionTrailWrapper>
          <ConversionTrail holdings={holdings} currencies={currencies} />
        </ConversionTrailWrapper>
      ) : null}
    </Wrapper>
  );
};

export default CurrencyConverter;
