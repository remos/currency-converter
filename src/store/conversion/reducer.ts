import { createSlice } from '@reduxjs/toolkit';
import { Holding } from 'types';
import { convert } from '../../data';
import { clearConversion, recalculateConversion } from './actions';

const initialState: {
  error?: string;
  holdings: Holding[];
} = {
  error: null,
  holdings: [],
};

export default createSlice({
  name: 'conversion',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      recalculateConversion,
      (state, { payload: { base, term, conversions } }) => {
        try {
          return {
            error: null,
            holdings: convert(base, term, conversions),
          };
        } catch (e) {
          return {
            error: e.toString(),
            holdings: [],
          };
        }
      }
    );

    builder.addCase(clearConversion, () => {
      return {
        error: null,
        holdings: [],
      };
    });
  },
}).reducer;
