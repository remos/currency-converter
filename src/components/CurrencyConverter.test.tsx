import { configureStore } from '@reduxjs/toolkit';
import React from 'react';
import { render } from '@testing-library/react';
import { RootState } from 'types';

import conversion from '../store/conversion/reducer';
import { Provider } from 'react-redux';
import { triggerRecalculateIfNeeded } from './CurrencyConverter';
import { clearConversion, recalculateConversion } from '../store/conversion/actions';

function wrappedRender(
  ui: React.ReactElement,
  {
    initialState,
    ...renderOptions
  }: {
    initialState: RootState;
  }
) {
  const store = configureStore({
    reducer: {
      conversion: conversion,
    },
    preloadedState: initialState,
  });

  const Wrapper: React.FC = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

describe('triggerRecalculateIfNeeded', () => {
  it('dispatches clear with invalid inputs', () => {
    const dispatch = jest.fn();
    triggerRecalculateIfNeeded(dispatch, '', 15, 'AUD', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());

    triggerRecalculateIfNeeded(dispatch, 'AUD', Number.NaN, 'AUD', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());

    triggerRecalculateIfNeeded(dispatch, 'AUD', 15, '', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());

    triggerRecalculateIfNeeded(dispatch, '', Number.NaN, '', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());

    triggerRecalculateIfNeeded(dispatch, 'AUD', Number.NaN, '', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());

    triggerRecalculateIfNeeded(dispatch, '', Number.NaN, 'AUD', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());

    triggerRecalculateIfNeeded(dispatch, '', 15, '', {});
    expect(dispatch).toHaveBeenLastCalledWith(clearConversion());
  });

  it('dispatches recalculation request with valid inputs', () => {
    const dispatch = jest.fn();
    triggerRecalculateIfNeeded(dispatch, 'AUD', 15, 'JPY', {});
    expect(dispatch).toHaveBeenLastCalledWith(
      recalculateConversion({
        base: {
          currency: 'AUD',
          amount: 15,
        },
        term: 'JPY',
        conversions: {},
      })
    );
  });
});
