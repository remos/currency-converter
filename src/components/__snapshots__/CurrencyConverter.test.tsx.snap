// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter Dropped decimals when changing base currency 1`] = `
.c0 {
  background-color: white;
  padding: 20px;
  border-radius: 4px;
  border: 1px solid #aaa;
}

.c1 {
  margin-right: 4px;
}

.c2 {
  margin-top: 1rem;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
    >
      <input
        aria-label="base amount"
        class="c1"
        name="base-amount"
        type="text"
        value="12.12"
      />
      <select
        aria-label="base currency"
        name="base-currency"
      >
        <option />
        <option
          value="AUD"
        >
          AUD
        </option>
        <option
          value="CAD"
        >
          CAD
        </option>
        <option
          value="CNY"
        >
          CNY
        </option>
        <option
          value="CZK"
        >
          CZK
        </option>
        <option
          value="DKK"
        >
          DKK
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="JPY"
        >
          JPY
        </option>
        <option
          value="NOK"
        >
          NOK
        </option>
        <option
          value="NZD"
        >
          NZD
        </option>
        <option
          value="USD"
        >
          USD
        </option>
      </select>
       to 
      <select
        aria-label="term currency"
        name="term-currency"
      >
        <option />
        <option
          value="AUD"
        >
          AUD
        </option>
        <option
          value="CAD"
        >
          CAD
        </option>
        <option
          value="CNY"
        >
          CNY
        </option>
        <option
          value="CZK"
        >
          CZK
        </option>
        <option
          value="DKK"
        >
          DKK
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="JPY"
        >
          JPY
        </option>
        <option
          value="NOK"
        >
          NOK
        </option>
        <option
          value="NZD"
        >
          NZD
        </option>
        <option
          value="USD"
        >
          USD
        </option>
      </select>
    </div>
    <div
      class="c2"
    >
      <div>
        <div>
          12.12 CAD
           =
           
          10.56 USD
        </div>
        <div>
          10.56 USD
           =
           
          8.57 EUR
        </div>
        <div>
          8.57 EUR
           =
           
          63.79 DKK
        </div>
        <hr />
        <div>
          12.12 CAD
           =
           
          63.79 DKK
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyConverter Unsanitary value 1`] = `
.c0 {
  background-color: white;
  padding: 20px;
  border-radius: 4px;
  border: 1px solid #aaa;
}

.c1 {
  margin-right: 4px;
}

.c2 {
  margin-top: 1rem;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
    >
      <input
        aria-label="base amount"
        class="c1"
        name="base-amount"
        type="text"
        value="12"
      />
      <select
        aria-label="base currency"
        name="base-currency"
      >
        <option />
        <option
          value="AUD"
        >
          AUD
        </option>
        <option
          value="CAD"
        >
          CAD
        </option>
        <option
          value="CNY"
        >
          CNY
        </option>
        <option
          value="CZK"
        >
          CZK
        </option>
        <option
          value="DKK"
        >
          DKK
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="JPY"
        >
          JPY
        </option>
        <option
          value="NOK"
        >
          NOK
        </option>
        <option
          value="NZD"
        >
          NZD
        </option>
        <option
          value="USD"
        >
          USD
        </option>
      </select>
       to 
      <select
        aria-label="term currency"
        name="term-currency"
      >
        <option />
        <option
          value="AUD"
        >
          AUD
        </option>
        <option
          value="CAD"
        >
          CAD
        </option>
        <option
          value="CNY"
        >
          CNY
        </option>
        <option
          value="CZK"
        >
          CZK
        </option>
        <option
          value="DKK"
        >
          DKK
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="JPY"
        >
          JPY
        </option>
        <option
          value="NOK"
        >
          NOK
        </option>
        <option
          value="NZD"
        >
          NZD
        </option>
        <option
          value="USD"
        >
          USD
        </option>
      </select>
    </div>
    <div
      class="c2"
    >
      <div>
        <div>
          12.00 CAD
           =
           
          10.45 USD
        </div>
        <div>
          10.45 USD
           =
           
          8.49 EUR
        </div>
        <div>
          8.49 EUR
           =
           
          63.16 DKK
        </div>
        <hr />
        <div>
          12.00 CAD
           =
           
          63.16 DKK
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyConverter Valid conversion 1`] = `
.c0 {
  background-color: white;
  padding: 20px;
  border-radius: 4px;
  border: 1px solid #aaa;
}

.c1 {
  margin-right: 4px;
}

.c2 {
  margin-top: 1rem;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
    >
      <input
        aria-label="base amount"
        class="c1"
        name="base-amount"
        type="text"
        value="12"
      />
      <select
        aria-label="base currency"
        name="base-currency"
      >
        <option />
        <option
          value="AUD"
        >
          AUD
        </option>
        <option
          value="CAD"
        >
          CAD
        </option>
        <option
          value="CNY"
        >
          CNY
        </option>
        <option
          value="CZK"
        >
          CZK
        </option>
        <option
          value="DKK"
        >
          DKK
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="JPY"
        >
          JPY
        </option>
        <option
          value="NOK"
        >
          NOK
        </option>
        <option
          value="NZD"
        >
          NZD
        </option>
        <option
          value="USD"
        >
          USD
        </option>
      </select>
       to 
      <select
        aria-label="term currency"
        name="term-currency"
      >
        <option />
        <option
          value="AUD"
        >
          AUD
        </option>
        <option
          value="CAD"
        >
          CAD
        </option>
        <option
          value="CNY"
        >
          CNY
        </option>
        <option
          value="CZK"
        >
          CZK
        </option>
        <option
          value="DKK"
        >
          DKK
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="JPY"
        >
          JPY
        </option>
        <option
          value="NOK"
        >
          NOK
        </option>
        <option
          value="NZD"
        >
          NZD
        </option>
        <option
          value="USD"
        >
          USD
        </option>
      </select>
    </div>
    <div
      class="c2"
    >
      <div>
        <div>
          12.00 CAD
           =
           
          10.45 USD
        </div>
        <div>
          10.45 USD
           =
           
          8.49 EUR
        </div>
        <div>
          8.49 EUR
           =
           
          63.16 DKK
        </div>
        <hr />
        <div>
          12.00 CAD
           =
           
          63.16 DKK
        </div>
      </div>
    </div>
  </div>
</div>
`;
